proposals/0163-lift-cpi-caller-restriction.md

```diff
@@ -0,0 +1,96 @@
+---
+simd: '0163'
+title: Lift the CPI caller restriction
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Draft
+created: 2024-07-16
+feature: TBD
+---
+
+## Summary
+
+Remove the check which forces CPI callers to have the program account of the
+callee available to them as an instruction account.
+
+## Motivation
+
+The restriction is purely historical and not necessary for the protocol or its
+implementations.
+
+Removing it would improve composability and reduce transaction building
+complexity because program accounts would no longer be passed all the way from
+transaction level instructions to the inner most CPI instructions.
+
+Additionally, not passing these program accounts is a significant reduction in
+data copied during serialization as program accounts of loader v1, v2 and v4
+contain the entire 10 MB ELF. This massively reduces the CU cost for nested CPI
+and disincentivizes programs from accessing program accounts, which could later
+allow us to remove program loading from transaction loading.
+
+### Example transaction
+
+- Accounts:
+  - Fee payer
+  - Caller Program
+  - Callee Program
+  - Token
+- Transaction-level instruction:
+  - Program Account: Caller Program
+  - Instruction Accounts:
+    - **Callee Program** (this could be removed)
+    - Token
+  - CPI:
+    - Program Account: Callee Program
+    - Instruction Accounts:
+      - Token
+
+Currently, in order to be used as a program account for a nested instruction
+the callee program must be passed as an instruction account to all outer
+instructions recursively.
```

Comment:
It's not immediately clear to me from reading this proposal how a user might add a callee program to the transaction account list. Currently, most people build transactions by providing a list of instructions, a fee payer, and a blockhash, then apply signatures. Programs which are targeted by instructions are sourced from each instruction's `program_id`.

How does a user provide the callee program ID to the transaction without adding it to any instruction? If it's hard-coded in the caller program - as mentioned below - how would the transaction know about it?

Reply:
That is a good point, but independent of the protocol or validator implementation. The ecosystem will have to figure out how they can profit from it, this change just enables them to do so and gives a few hints and ideas about how it could be done.

Reply:
Okay, that's fair, but what about the transaction-level protocol? Does it matter which index in the transactions accounts this callee program exists at? What about for multiple callee programs?

Reply:
Except for the fee payer which must stay in index 0, all other transaction accounts order is irrelevant. But we could add wording that we suggest to add them at the end.

Reply:
Yeah, maybe just in the suggestion bit at the end?